shared/stack/stack_service.go

```diff
@@ -0,0 +1,430 @@
+	if dryRun {
+		log.Debugf("temporarily creating a TFE workspace for stack '%s'", stackName)
+	} else {
+		log.Debugf("creating stack '%s'", stackName)
+	}
+
+	enableDynamoLocking, ok := ctx.Value(options.EnableDynamoLockingKey).(bool)
+	if !ok {
+		enableDynamoLocking = false
+	}
+	var err error
+	if enableDynamoLocking {
+		err = s.addToStacklistWithLock(ctx, stackName)
+	} else {
+		err = s.addToStacklist(ctx, stackName)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if !util.IsLocalstackMode() {
+		// Create the workspace
+		if err := s.resync(ctx, opts...); err != nil {
+			return nil, err
+		}
+	}
+
+	_, err = s.GetStackWorkspace(ctx, stackName)
+	if err != nil {
+		return nil, err
+	}
+	return s.createStack(stackName), nil
+}
+
+func (s *StackService) addToStacklistWithLock(ctx context.Context, stackName string) error {
+	log.WithField("stack_name", stackName).Debug("Adding new stack with a lock...")
+	distributedLock, err := s.getDistributedLock()
+	if err != nil {
+		return err
+	}
+	defer distributedLock.Close(ctx)
+
+	lockKey := s.GetNamespacedWritePath()
+	lock, err := distributedLock.AcquireLock(ctx, lockKey)
+	if err != nil {
+		return err
+	}
+
+	// don't return if there was an error here, we still need to release the lock so we'll use multierror instead
+	ret := s.addToStacklist(ctx, stackName)
+
+	_, err = distributedLock.ReleaseLock(ctx, lock)
+	if err != nil {
+		ret = multierror.Append(ret, errors.Wrapf(err, "unable to release the lock on %s", lockKey))
+	}
+
+	return ret
+}
+
+func (s *StackService) addToStacklist(ctx context.Context, stackName string) error {
+	log.WithField("stack_name", stackName).Debug("Adding new stack...")
+	existStacks, err := s.GetStacks(ctx)
+	if err != nil {
+		return err
+	}
+
+	newStackNames := []string{}
+	stackNameExists := false
+	for name := range existStacks {
+		newStackNames = append(newStackNames, name)
+		if name == stackName {
+			stackNameExists = true
+		}
+	}
+	if !stackNameExists {
+		newStackNames = append(newStackNames, stackName)
+	}
+
+	return s.writeStacklist(ctx, newStackNames)
+}
+
+func (s *StackService) writeStacklist(ctx context.Context, stackNames []string) error {
+	sort.Strings(stackNames)
+
+	stackNamesJson, err := json.Marshal(stackNames)
+	if err != nil {
+		return errors.Wrap(err, "unable to serialize stack list as json")
+	}
+
+	stackNamesStr := string(stackNamesJson)
+	log.WithFields(log.Fields{"path": s.GetNamespacedWritePath(), "data": stackNamesStr}).Debug("Writing to paramstore...")
+	cb, err := s.backend.GetComputeBackend(ctx)
+	if err != nil {
+		return errors.Wrap(err, "unable to connect to a compute backend")
+	}
+
+	if err := cb.WriteParam(ctx, s.GetNamespacedWritePath(), stackNamesStr); err != nil {
+		return errors.Wrap(err, "unable to write a workspace param")
+	}
+	log.WithFields(log.Fields{"path": s.GetWritePath(), "data": stackNamesStr}).Debug("Writing to paramstore...")
+	if err := cb.WriteParam(ctx, s.GetWritePath(), stackNamesStr); err != nil {
```

Comment:
Which one of these two params is the source of truth?

Reply:
all this code is copied from the CLI. Do you want me to reimplement it in this PR?